### PR TITLE
CARv2 read support (pending final specification of CARv2)

### PR DESCRIPTION
**Pull request by rvagg**
Adds the ability to read a CAR that is in CARv2 format as it's currently defined. The format is not final so this PR won't be merged until it is. Current work is going on in https://github.com/ipld/go-car `v2` directory and I need to make a spec for this for further discussion.

Format details can be seen here in the adjusted `readHeader()` and `readV2Header()` which is the primary difference. CARv2 is a pragmatic format that simply wraps a CARv1, appending a 11 byte series of magic bytes and a header comprised of fixed-length uint64's describing capabilities as well as where the CARv1 is located and where a trailing index is located (if there is one).

This implementation ignores any trailing index and simply reads the CAR as normal, transparently. The header will be different because it adds some new fields but still has `version` and `roots`. So it should be a non-breaking feature upgrade as it is now.

### CODE REVIEW

File: lib/decoder.js [hunk 1]
```diff
@@ -33,29 +38,60 @@ async function readVarint (reader) {
...
-  if (Object.keys(block).filter((p) => p !== 'roots' && p !== 'version').length) {
-    throw new Error('Invalid CAR header format')
+  if (block.version === 1) {
+    return block
   }
-  return block
+  // version 2
+  const v2Header = await readV2Header(reader)
+  reader.seek(reader.pos - v2Header.dataOffset)
+  const v1Header = await readHeader(reader, 1)
```
Comment (willscott) [inline lib/decoder.js:93]: could this lead to recursive weirdness where you put a v2 in a v2?

Comment (rvagg) [reply lib/decoder.js:93]: shouldn't be, that's what the optional second argument is for in this case - the `block.version !== strictVersion` should mean that if it's not a v1 header then it'll error before it even gets down here on the recursive call
not a bad idea for a test though

Comment (rvagg) [reply lib/decoder.js:93]: added test for failing a recursive carv2 @ ad8cdc0
good point to make explicit in the spec I think

**Comment by rvagg** [comment]
need to verify this against the current go-car output before it can be merged

**Comment by rvagg** [comment]
@biglep yep, I don't think it's a big job, I also don't think this is particularly urgent since there's nobody outside of Lotus that's actually touching CARv2 files, certainly not in JS. Will get this done soon though.

**Comment by BigLep** [comment]
Per slack: "this is not critical work as we don’t expect CARv2 to see much (or any) JS use for now, but it’s close to being done, I just need to do some more verification that the final CARv2 format matches the work I did there (which was pre-final, assisting spec writing)."

**Comment by rvagg** [comment]
I'm labelling this as "help wanted" and "exp/beginner" because it's mostly done and just needs a bit of love to push it over the line and get some proper tests in to confirm it works. go-car now has some decent test fixture .car files that could be re-used here to confirm that it works. We're not supporting _reading_ the index, just loading a CARv2 as if it's a CARv1 for now, so this isn't very hard.

I'll get back to this sometime, but someone else is welcome to help out if they want! Ping me if you need help.

**Comment by rvagg** [comment]
Okie dokie, this is good to go.

* Adds transparent CARv2 read support, so it'll handle a CARv2 as if it's a CARv1, the only way you'll know is if you interrogate the `version` property of the reader
* No write support
* No index support (read/use or write)
* Stole the go-car fixture CARs and embedded them here, they're quite large and I've base64'd them into .js files so they can easily run in the browser tests. I've also had to exclude them from the published bundle, so the tests are no longer complete (not that anyone will complain, it's common to have complaints that the tests are _in_ the bundle at all)
  - https://github.com/ipld/go-car/pull/293 is an adjustment to the fixtures that I'd like to push back to go-car, depending on the outcome of that I may have to adjust here

Future work on CARv2 in JS would be focused on server-side functionality I think, there's not a good case for writing CARv2 in the browser and the index doesn't help much in the browser either. Being able to use a CAR as a blockstore on the server would be a nice feature IMO, but a low priority for now I think.

**Comment by rvagg** [comment]
I've changed this to a semver-major release because there's a type change to the `header` property (and `readHeader()` function on decoder if anyone's using that. It now returns a `Promise<CarHeader|CarV2Header>` instead of just a `Promise<CarHeader>` which may cause some grief even though `CarV2Header` is mostly an extension of `CarHeader`.

**Pull request merged**